subsys/storage/storage_area/storage_area_store.c

```diff
@@ -0,0 +1,1097 @@
+{
+	const struct storage_area *area = record->store->area;
+	const size_t crc_skip = record->store->crc_skip;
+	const size_t recpos = record->sector * record->store->sector_size +
+			      record->loc + SAS_HDRSIZE + crc_skip;
+	uint32_t crc = SAS_CRCINIT;
+	uint8_t buf[SAS_MAX(SAS_MINBUFSIZE, area->write_size)];
+	struct storage_area_iovec rd = {
+		.data = &buf,
+	};
+	size_t rdlen = record->size - crc_skip;
+	sa_off_t rdoff = recpos;
+
+	while (rdlen != 0U) {
+		rd.len = SAS_MIN(sizeof(buf), rdlen);
+		if (storage_area_readv(area, rdoff, &rd, 1U) != 0) {
+			goto end;
+		}
+
+		crc = crc32_ieee_update(crc, buf, rd.len);
+		rdlen -= rd.len;
+		rdoff += rd.len;
+	}
+
+	rd.len = SAS_CRCSIZE;
+	if (storage_area_readv(area, rdoff, &rd, 1U) != 0) {
+		goto end;
+	}
+
+	if (crc != sys_get_le32(buf)) {
+		goto end;
+	}
+
+	return true;
+end:
+	LOG_DBG("record at [%d-%d] has bad crc", record->sector, record->loc);
+	return false;
+}
+
+static int store_record_next_in_sector(struct storage_area_record *record,
+				       bool wrapcheck, bool recover)
+{
+	const struct storage_area_store *store = record->store;
+	const struct storage_area_store_data *data = store->data;
+	const struct storage_area *area = store->area;
+	const size_t secpos = record->sector * store->sector_size;
+	bool crc_ok = true;
+	int rc = -ENOENT;
+
+	if ((record->loc == 0U) && (store->sector_cookie != NULL) &&
+	    (store->sector_cookie_size != 0U)) {
+		const size_t cksz = store->sector_cookie_size;
+
+		record->loc = SAS_ALIGNUP(cksz, area->write_size);
+	}
+
+	while (true) {
+		uint8_t buf[SAS_HDRSIZE];
+		struct storage_area_iovec rd = {
+			.data = &buf,
+			.len = sizeof(buf),
+		};
+		size_t rdpos = record->loc;
+		sa_off_t rdoff;
+
+		if (record->size != 0U) {
+			rdpos += (SAS_HDRSIZE + record->size + SAS_CRCSIZE);
+			rdpos = SAS_ALIGNUP(rdpos, area->write_size);
+		}
+
+		if (((data->sector == record->sector) && (data->loc <= rdpos)) ||
+		    (rdpos == store->sector_size)) {
+			record->loc = rdpos;
+			record->size = 0U;
+			rc = -ENOENT;
+			break;
+		}
+
+		rdoff = secpos + rdpos;
+		rc = storage_area_readv(area, rdoff, &rd, 1U);
+		if (rc != 0) {
+			break;
+		}
+
+		size_t rsize = (size_t)sys_get_le16(&buf[2]);
+		size_t avail =
+			store->sector_size - rdpos - SAS_CRCSIZE - SAS_HDRSIZE;
+		bool size_ok = ((rsize > 0U) && (rsize < avail));
+
+		if (record->sector > data->sector) {
+			buf[1]++;
+		}
+
+		if (!wrapcheck) {
+			buf[1] = data->wrapcnt;
+		}
+
+		if ((size_ok) && (!crc_ok)) {
+			record->loc = rdpos;
+			record->size = rsize;
+			crc_ok = store_record_valid(record);
+		}
+
+		if ((buf[0] == SAS_MAGIC) && (buf[1] == data->wrapcnt) &&
+		    (size_ok) && (crc_ok)) {
+			record->loc = rdpos;
+			record->size = rsize;
+			rc = 0;
+			break;
+		}
+
+		if (!recover) {
+			rc = -ENOENT;
+			break;
+		}
+
+		crc_ok = false;
+		record->loc += area->write_size;
+		record->size = 0U;
+	}
+
+	if ((rc != 0) && (rc != -ENOENT)) {
```

Comment:
There is no flash, eeprom or disk driver that returns -ENOENT, but I will change it. BTW this would also be a problem for nvs.

Reply:
fixed

Reply:
> There is no flash, eeprom or disk driver that returns -ENOENT, but I will change it. BTW this would also be a problem for nvs.

Maybe then it should be reported as bug to fix.